### PR TITLE
Add ViT Vision Model Support

### PR DESCRIPTION
**Pull request by jacobbieker**
This PR is a subset of #3 that is only the changes to the `Model` class for supporting ranked image reward training. 

These image models are primarily designed to work with the dataset from #5.

There are also some unit tests included for testing the loading of the model and forward passes.

### CODE REVIEW

File: src/model.py [hunk 1]
```diff
@@ -3,62 +3,134 @@
...
+        """
         super().__init__()
+        self.vision_model = vision_model
         self.config = transformer.config
-        self.config.n_embd = self.config.hidden_size if hasattr(self.config, "hidden_size") else self.config.n_embd
+        self.config.n_embd = (
+            self.config.hidden_size if hasattr(self.config, "hidden_size") else self.config.n_embd
+        )
         self.transformer = transformer
         self.v_head = nn.Linear(self.config.n_embd, 1, bias=False)
```
Comment (amydeng2000) [inline src/model.py:22]: @UYasher and I briefly talked about this, I don't think there's any benefits in setting bias=False. What do you think Jacob? Should we just remove it?

Comment (jacobbieker) [reply src/model.py:22]: Yeah, I don't really see the benefits either, so yeah, I cna remove it

File: src/model.py [hunk 2]
```diff
@@ -3,62 +3,134 @@
...
         self.config.torch_dtype = next(transformer.parameters()).dtype
         self.v_head = self.v_head.to(dtype=self.config.torch_dtype)
         self.tokenizer = tokenizer
-        self.PAD_ID = tokenizer.convert_tokens_to_ids(tokenizer.pad_token)
+        if hasattr(tokenizer, "pad_token"):
+            self.PAD_ID = tokenizer.convert_tokens_to_ids(tokenizer.pad_token)
         self.max_ranks_per_batch = max_ranks_per_batch
 
-
     def forward(
```
Comment (amydeng2000) [inline src/model.py:30]: I had a working version of this earlier (without slicing and averaging) and it looked just like what you have here, so I am pretty sure everything here is good. However, I have forgotten about how I wrote my code lol. Can you remind me what the following variables stand for / do? I gave my best effort explanation, but please correct me.

- `input_data` should have a shape of (num_websites * max_ranks_per_batch) x 1 x 3(rgb channels) x img_width(defualt 224) x img_height(default 224)
- `bs`: is bs == len(counts)?
- `ranked_rewards` is of length `max_ranks_per_batch`. What does `ranked_rewards[i]` represent here?
- `end_scores` have length `max_ranks_per_batch` as well, but I am also confused on what `end_scores[i]` represents. Is it a list of scores (calculated by averaging reward tokens) for the 1st ranking year of each website?

File: src/utils.py
Comment (amydeng2000) [inline src/utils.py:1]: I think this file is serving PR #7 instead of this PR since none of the methods are used in code in this branch

File: src/model.py (hunk 2 above)
Comment (amydeng2000) [reply src/model.py:30]: also since we changed the max_ranks_per_batch in the previous PR, let's change all references to 27 in this file and test_model.py as well

File: src/utils.py
Comment (amydeng2000) [reply src/utils.py:1]: If I am right, then let's remove the references to the functions in __init__ as well for this PR

File: src/model.py (hunk 2 above)
Comment (jacobbieker) [reply src/model.py:30]: 1. ```input_data``` has shape of  ```[Num_websites*max_ranks_per_batch+num_extra_slices, 3 (RGB), Width(224), Height(224)]``` where ```num_extra_slices``` is the number of extra image slices. In essence, the first dimension of ```input_data``` would be the sum of all the list of image slice counts in ```counts```. This means there is at least ```num_websites * max_ranks_per_batch``` but up to a max of ```num_websites*max_ranks_per_batch*max_slices_per_image``` 
2. ```bs```: This is ```num_websites```, as after lines 66-80, where the multiple image slice embeddings are averaged and combined, ```input_data``` will always have shape ```[Num_websites*max_ranks_per_batch, 3 (RGB), Width(224), Height(224)]```, so dividing by the ```max_ranks_per_batch``` on line 83 then makes ```bs = num_websites``` 
3. ```ranked_rewards[i]``` is the rewards for that rank across all batches in the inputs. Writing this out showed a bug that is now fixed, so as it stands with that bug fixed, ```ranked_rewards[0]``` would be the hidden states for ```bs``` 2023 websites, which would all be ranked 1st. 
4. Yes, that is correct, ```end_scores``` averages across all ```num_websites``` for the 1st ranking year of each website. It is the net score for that rank in that batch, which is why it is length ```max_ranks_per_batch```.

Thanks for these questions! Explaining it made me find a bug in how I was doing it.

File: src/utils.py
Comment (jacobbieker) [reply src/utils.py:1]: Yep, thanks!

File: src/model.py [hunk 3]
```diff
@@ -3,62 +3,156 @@
...
+                    slice_count += slice_count_lst[i]
+                # To remake batch size, need to concatenate again
+                averaged_rewards.append(torch.cat(website_rewards))
+                averaged_input_data.append(torch.cat(website_images))
+            rewards = torch.cat(averaged_rewards)
+            # Input data is now average of all the slices, so does not make much physical sense, but we need it for
+            # knowing where the padding is
+            input_data = torch.cat(averaged_input_data)
+            # The 'ranked' and 'ranked_rewards' code below assumes that for the given batch size, the highest ranked
+            # element in each batch is at the front, i.e. the data is structured
```
Comment (jacobbieker) [inline src/model.py:93]: I'm thinking I should move this to the dataset, to ensure the input images are in the right order. This seems to add a decent amount of overhead when I'm trying to train a model. The averaging still needs to happen inside ```Model``` but the reordering can happen in the ```ImageCollator```

Comment (amydeng2000) [reply src/model.py]: that makes sense

File: src/model.py [hunk 4]
```diff
@@ -78,19 +167,25 @@ def forward(
...
+            end_of_prompt = self.get_end_of_prompt(input1, input2)
+            start_of_padding = max(
+                self.get_start_of_padding(input1), self.get_start_of_padding(input2)
+            )
+            return -torch.log(
+                torch.sigmoid(
+                    reward1[end_of_prompt:start_of_padding]
+                    - reward2[end_of_prompt:start_of_padding]
+                )
+            ).mean()
```
Comment (UYasher) [inline src/model.py:188]: This is mostly a note to myself, but I think this difference is caused by using a decoder-based model in the text case (either that, or using the hidden states). With an encoder-only model we could get a single value by simply running the model, I think, instead of getting a value for each token. It's not immediately obvious to me which method is better.

File: tests/test_model.py
Comment (UYasher) [inline tests/test_model.py:1]: Curious how you came up with these tests, and what best-practices there are around writing tests for models. I typically haven't written tests (but that's because I come from a research background, so code was less-than-production-grade).

**Review by UYasher — APPROVED** [review]

File: src/model.py (hunk 2 above)
Comment (jacobbieker) [reply src/model.py:30]: Yes, it is, depending on how long the screenshot is, websites will have between 1 and `max_slices` images for a given year.

File: tests/test_model.py
Comment (jacobbieker) [reply tests/test_model.py:1]: I'm not sure the best practices around tests for models, and some of these, like loading Llama-7B, probably take too long to really be unit tests, but I wrote these to make sure that I can relatively quickly (especially locally once the models are cached) ensure that the models 1. load, including a variety of different models as this code shouldn't be tied to a single model, 2. can take data in the shapes we expect without outputting NaNs, especially when there is no NaNs or infs in the inputs, 3. that the loss and other outputs work as we expect them to, having the correct shapes and the loss is valid. It is also faster and easier to check if these tests pass to see if my internal changes to `Model`, especially around the averaging of multiple slices per year, was working, rather than trying to train a model and wait for the startup time for training. 

Generally, for unit testing models, from my experience its more about making sure the model you are using is not introducing NaNs or inf's, or some other error in the architecture itself. This way you can more be sure that an error you encounter is most likely because of the data pipeline and some potential problem there, rather than the model. With HF models, the model is mostly all one thing, but for example, I wrote an implementation of a Deepmind model, and [tested all the individual components separately, and together](https://github.com/openclimatefix/skillful_nowcasting/blob/main/tests/test_model.py).
